### DISCUSSION TITLE
SOAP to REST transformation

### DISCUSSION THREAD

**Discussion by sonicaghi** [Kong/kong#7359] [Ideas and feature requests]

**Comment by lucamaraschi** [comment]
:+1:

**Comment by kyyberi** [comment]
Might not be out of the box solution, but there's one implementation for REST-SOAP-REST conversion feature (in Estonian X-Road) created by Finnish developers. https://github.com/educloudalliance/xroad-rest-gateway

**Comment by iopsthecloud** [comment]
:+1:

**Comment by AlexMikhalev** [comment]
I am looking forward for it to become part of the kong solution!

**Comment by BrianHutchison** [comment]
Is it too late to +1 this? 

Would be nice to specify apikey in wsse:Security, e.g. 

``` xml
<soapenv:Header>
                        <wsse:Security>
                                    <wsse:UsernameToken>
                                                <wsse:Username>fe7060ae-2907-4f90-9d60-f0ba9d9ec5e4</wsse:Username>
                                    </wsse:UsernameToken>
                        </wsse:Security>
</soapenv:Header>
```

**Comment by willfarrell** [comment]
lua starting point: https://github.com/tomasguisasola/luasoap

nodejs implementation by IBM: https://github.com/strongloop/loopback-connector-soap

**Comment by mdan1eli** [comment]
I would really start from a SOAP authentication plugin and keep Kong more as a gateway for exposure, more than an integration layer for consumption.  We have camel routes runnning in containers behind Kong for that, though i know that technically a plugin can be built.  Don't want functional stuff on this layer. I want policies only.

**Comment by AlexMikhalev** [comment]
Having it via plugin would be fine. Also that's why apigee bought by Google - this is important functionality for gateway to have to be able to serve in enterprise world.

**Comment by mdan1eli** [comment]
SOAP plugins for authentication (username, certificate and saml), backend protection (payload size limiting and payload validation) and monitoring (logging and auditing). They are very similar to stuff that already exists for generic HTTP services just a bit specialized.  The soap to rest can also be seen as a specialization of the transformation plugins, but in truth soap-rest is a bit more than xml <> json to me.  For example support for ws security and ws trust standards (how to map these standards in the transformation) and it gets more complicated and that is why i separate the two needsand  I want to keep my gateway simple to operate and avoid being over ambitious).  Having above soap capabilities as plugin would extend reach to scenarios where we employ tools like IBM DP as a SOA Gateway for B2B.

**Comment by sonicaghi** [comment]
Some discussed here https://github.com/Mashape/kong/issues/1536

**Comment by willfarrell** [comment]
SOAP plugins for authentication (applying digital signatures and encrypt/decrypt transformations) would be a great first step and be better suited to the scope of kong as @mdan1eli mentioned. I just implemented something very similar in node using https://github.com/yaronn/ws.js. @yaronn might be a good technical resource on this piece and would definitely be able to scope out the requirements needed to tackle this.

**Comment by arun-c-g** [comment]
Is SOAP to REST plugin available in the Enterprise subscription

**Comment by puttareddy** [comment]
+1 .

**Comment by mihir83in** [comment]
+1

**Comment by stephanstapel** [comment]
+1

**Comment by messi-gao** [comment]
+1

**Comment by momo2366** [comment]
+1

**Comment by skokaina** [comment]
+10

**Comment by vsadilov** [comment]
+1
